### PR TITLE
Optional local address

### PR DESCRIPTION
**Pull request by Infinisil**
This changes the type of `daLocalAddress` from `Either Socket FilePath` to `Maybe (Either Socket FilePath)`, allowing having no local address at all, which is useful if there's no need for node-to-client communication.

I also made a refactoring of the `withAsync` calls to remove the nesting, using [`Async.Concurrently`](https://hackage.haskell.org/package/async-2.2.3/docs/Control-Concurrent-Async.html#t:Concurrently)'s `Alternative` instance, which runs threads concurrently, waiting for the first result.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs [hunk 1]
```diff
@@ -203,40 +204,29 @@ runDataDiffusion tracers
...
-    void $
-      -- clean state thread
-      Async.withAsync (cleanNetworkMutableState networkState) $ \cleanNetworkStateThread ->
-
-        -- clean local state thread
-        Async.withAsync (cleanNetworkMutableState networkLocalState) $ \cleanLocalNetworkStateThread ->
-
+    let dnsSubActions = runDnsSubscriptionWorker snocket networkState lias <$> daDnsProducers
+        serverActions = case daDiffusionMode of
+          InitiatorAndResponderDiffusionMode -> runServer snocket networkState . fmap Socket.addrAddress <$> addresses
```
[the remark this answers is not in the document]
Comment (Infinisil) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs]: Done that

**Review by coot — APPROVED** [review]
LGTM

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Infinisil) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs]: (whoops, accidentally clicked :confused: instead of :+1:, fixed now)

**Comment by Infinisil** [comment]
Anything preventing a merge?

**Comment by coot** [comment]
I'll merge this tomorrow, I'm just postponing a merge conflict in our `p2p-master` branch.

**Comment by coot** [comment]
bors merge

**Pull request merged**